### PR TITLE
refactor(web): add codegen and change all api call to use codegen

### PR DESCRIPTION
**Pull request by Leomotors**
### NOTE: This PR merges into `dev` channel
Part of #371

[Read this](https://github.com/thinc-org/cugetreg/issues/371#issuecomment-1242659476)

(Recap)
In this ticket, I would like to split PR into 2 (or more parts) for easier reviewing.
As two PR are entirely independent, we can merge by part.

## Motivation and Changes
- Currently, we have two sets of GraphQL schema/types, one is in backend, one is in frontend
- We should have a Single Source of Truth. Like every company will do, we will use backend graphql as the main schema
- We will have codegen (`libs/codegen`) that reads the backend's graphql schema and generate types/documents/react hooks
- All apollo queries on the frontend should now all import types/documents/react hooks from `libs/codegen`

### Backend
- `Semester` type is different from the backend and requires backend changes
- `Semester` uses 1, 2, 3 as type making it not noice with enum. We will need to change in to First, Second, Third.

## Todo
- Remove `@thinc-org/chula-courses`, all usages have been removed from Frontend but not Backend.

### CODE REVIEW

**Comment by bombnp** [comment]
please tell us specifically what files to review

we are not reviewing 96 files

File: libs/codegen/src/future/index.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/**
+ * `@cugetreg/codegen/future` are types that we use
+ * but does not available yet in backend's GraphQL.
+ *
+ * TODO: Wait for backend to implement these types into GraphQL
+ */
+
+export enum Semester {
```
Comment (Nacnano) [inline libs/codegen/src/future/index.ts:8]: ? There is no need to add future/index.ts because, in the end, we will probably use it anyway.

**Review by Nacnano — CHANGES_REQUESTED** [review]

File: libs/codegen/src/future/index.ts (hunk 1 above)
Comment (Leomotors) [reply libs/codegen/src/future/index.ts:8]: temporary

Comment (bombnp) [reply libs/codegen/src/future/index.ts:8]: u could just add Semester enum in graphql right now btw in course.enum.graphql in cugetreg-api

**Comment by bombnp** [comment]
is this still draft?

**Comment by Leomotors** [comment]
@bombnp Will need to figure out how to deal with `Semester`, and it should now be ready to merge.

`Semester` type is "1" | "2" | "3" and we want to type it from backend but not sure if number will work with enum.

**Comment by bombnp** [comment]
> @bombnp Will need to figure out how to deal with `Semester`, and it should now be ready to merge.
> 
> `Semester` type is "1" | "2" | "3" and we want to type it from backend but not sure if number will work with enum.

just use string enums, all systems expected semester type to be strings (as intended, as it has no numerical meaning)

**Comment by Leomotors** [comment]
@bombnp We cannot do
```
enum Semester {
  1
  2
  3
}
```

We may need to change it to First, Second, Third which require backend changes.

**Comment by bombnp** [comment]
Changing to First, Second, Third enum would break our system in SO many places. Just use normal strings for now i guess.

I'll do a full review later tonight. Im a outside.

File: tsconfig.base.json [hunk 2]
```diff
@@ -16,6 +16,9 @@
     "baseUrl": ".",
     "paths": {
       "@api/*": ["apps/cugetreg-api/src/*"],
+      "@cugetreg/codegen": ["libs/codegen/src"],
+      "@cugetreg/codegen/future": ["libs/codegen/src/future"],
```
Comment (bombnp) [inline tsconfig.base.json:20]: do we need to create a separate path just for future tho?

File: libs/codegen/src/index.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+export * from './generated'
```
Comment (bombnp) [inline libs/codegen/src/index.ts:1]: please add a wiki with details about how to use codegen and stuffs

**Review by bombnp — APPROVED** [review]
god leomotor strikes again

few comments though

File: tsconfig.base.json (hunk 2 above)
Comment (Leomotors) [reply tsconfig.base.json:20]: We should separate because `future` is temporary.

Comment (saenyakorn) [inline tsconfig.base.json:20]: It would be great if we named the alias path `@lib/something`

Comment (paphonb) [reply tsconfig.base.json:20]: I agree. Something like `@libs/codegen` would be much easier to use.

Comment (bombnp) [reply tsconfig.base.json:20]: but it behaves the exact same way, no? What benefits does separating them gives? When we remove future, we'll have to remove `@cugetreg/codegen/future` anyways

**Pull request merged**